File: App_Flutter_rentalhouse/lib/views/Admin/Service/news.dart
// lib/services/news_service.dart

import 'dart:io';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter_rentalhouse/config/api_routes.dart';

class NewsService {
  final FirebaseAuth _auth = FirebaseAuth.instance;

  // ====================== HỖ TRỢ ======================

  /// Tự động lấy Firebase ID Token
  Future<String?> _getIdToken() async {
    final user = _auth.currentUser;
    if (user == null) return null;
    try {
      return await user.getIdToken();
    } catch (e) {
      print('Lỗi lấy Firebase ID Token: $e');
      return null;
    }
  }

  /// Tạo header có Authorization (nếu có token)
  Future<Map<String, String>> _headers({bool requireAuth = false}) async {
    final token = await _getIdToken();
    final headers = {
      'Content-Type': 'application/json',
    };

    if (token != null) {
      headers['Authorization'] = 'Bearer $token';
    } else if (requireAuth) {
      throw Exception('Chưa đăng nhập');
    }

    return headers;
  }

  /// Xác định MIME type từ đường dẫn file
  String _determineMimeType(String path) {
    final lower = path.toLowerCase();
    if (lower.endsWith('.png')) return 'image/png';
    if (lower.endsWith('.webp')) return 'image/webp';
    if (lower.endsWith('.jpg') || lower.endsWith('.jpeg')) return 'image/jpeg';
    return 'image/jpeg';
  }

  /// Kiểm tra kích thước file (max 10MB)
  Future<void> _validateImageFile(File file) async {
    if (!await file.exists()) {
      throw Exception('File ảnh không tồn tại');
    }
    final size = await file.length();
    const maxSize = 10 * 1024 * 1024; // 10MB
    if (size > maxSize) {
      throw Exception(
          'Ảnh quá lớn (tối đa 10MB, hiện tại ${(size / 1024 / 1024).toStringAsFixed(1)}MB)');
    }
  }

  // ====================== PUBLIC APIs ======================

  /// Lấy tin tức công khai (có phân trang)
  Future<Map<String, dynamic>> fetchAllNews({
    int page = 1,
    int limit = 10,
  }) async {
    final uri = Uri.parse('${ApiRoutes.news}?page=$page&limit=$limit');
    final response = await http.get(uri, headers: await _headers()).timeout(
      const Duration(seconds: 30),
      onTimeout: () => throw Exception('Timeout'),
    );

    if (response.statusCode == 200) {
      return jsonDecode(utf8.decode(response.bodyBytes));
    }
    throw Exception('Tải tin tức thất bại (${response.statusCode})');
  }

  /// Lấy tin tức nổi bật
  Future<List<dynamic>> fetchFeaturedNews({int limit = 3}) async {
    final uri = Uri.parse('${ApiRoutes.featuredNews}?limit=$limit');
    final response = await http.get(uri, headers: await _headers()).timeout(
      const Duration(seconds: 30),
      onTimeout: () => throw Exception('Timeout'),
    );

    if (response.statusCode == 200) {
      return jsonDecode(utf8.decode(response.bodyBytes)) as List;
    }
    throw Exception('Tải tin tức nổi bật thất bại');
  }

  /// Lấy chi tiết 1 tin tức (tăng view)
  Future<Map<String, dynamic>> fetchNewsDetail(String newsId) async {
    final uri = Uri.parse(ApiRoutes.newsById(newsId));
    final response = await http.get(uri, headers: await _headers()).timeout(
      const Duration(seconds: 30),
      onTimeout: () => throw Exception('Timeout'),
    );

    if (response.statusCode == 200) {
      return jsonDecode(utf8.decode(response.bodyBytes));
    } else if (response.statusCode == 404) {
      throw Exception('Tin tức không tồn tại');
    }
    throw Exception('Lấy chi tiết thất bại');
  }

  // ====================== SAVED ARTICLES APIs ======================

  /// Lưu tin tức
  Future<void> saveArticle(String newsId) async {
    try {
      final uri = Uri.parse(ApiRoutes.saveArticle(newsId));
      print('Saving article: $uri');

      final headers = await _headers(requireAuth: true);
      final response = await http
          .post(
        uri,
        headers: headers,
      )
          .timeout(
        const Duration(seconds: 30),
        onTimeout: () => throw Exception('Timeout'),
      );

      print('Save response status: ${response.statusCode}');

      if (response.statusCode == 201) {
        return;
      } else if (response.statusCode == 400) {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Bạn đã lưu tin tức này');
      } else if (response.statusCode == 401) {
        throw Exception('Token hết hạn, vui lòng đăng nhập lại');
      } else if (response.statusCode == 404) {
        throw Exception('Tin tức không tồn tại');
      }
      throw Exception('Không thể lưu tin tức (${response.statusCode})');
    } catch (e) {
      print('Error in saveArticle: $e');
      rethrow;
    }
  }

  /// Bỏ lưu tin tức
  Future<void> unsaveArticle(String newsId) async {
    try {
      final uri = Uri.parse(ApiRoutes.unsaveArticle(newsId));
      print('Unsaving article: $uri');

      final headers = await _headers(requireAuth: true);
      final response = await http
          .delete(
        uri,
        headers: headers,
      )
          .timeout(
        const Duration(seconds: 30),
        onTimeout: () => throw Exception('Timeout'),
      );

      if (response.statusCode == 200) {
        return;
      } else if (response.statusCode == 401) {
        throw Exception('Token hết hạn');
      } else if (response.statusCode == 404) {
        throw Exception('Tin tức không tồn tại');
      }
      throw Exception('Không thể bỏ lưu tin tức (${response.statusCode})');
    } catch (e) {
      print('Error in unsaveArticle: $e');
      rethrow;
    }
  }

  /// Kiểm tra tin tức có được lưu không
  Future<bool> checkIsSaved(String newsId) async {
    try {
      final uri = Uri.parse(ApiRoutes.checkIsSaved(newsId));
      final headers = await _headers(requireAuth: true);
      final response = await http
          .get(
        uri,
        headers: headers,
      )
          .timeout(
        const Duration(seconds: 30),
        onTimeout: () => throw Exception('Timeout'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['isSaved'] ?? false;
      }
      return false;
    } catch (e) {
      print('Error in checkIsSaved: $e');
      return false;
    }
  }

  /// Lấy tất cả tin tức đã lưu của user (có phân trang)
  Future<Map<String, dynamic>> fetchSavedArticles({
    int page = 1,
    int limit = 10,
  }) async {
    try {
      final uri =
      Uri.parse('${ApiRoutes.savedArticles}?page=$page&limit=$limit');
      print('Fetching saved articles: $uri');

      final headers = await _headers(requireAuth: true);
      final response = await http
          .get(
        uri,
        headers: headers,
      )
          .timeout(
        const Duration(seconds: 30),
        onTimeout: () => throw Exception('Timeout'),
      );

      if (response.statusCode == 200) {
        return jsonDecode(utf8.decode(response.bodyBytes));
      } else if (response.statusCode == 401) {
        throw Exception('Token hết hạn, vui lòng đăng nhập lại');
      }
      throw Exception('Tải tin tức đã lưu thất bại (${response.statusCode})');
    } catch (e) {
      print('Error in fetchSavedArticles: $e');
      rethrow;
    }
  }

  // ====================== ADMIN APIs ======================

  /// Admin: Lấy tất cả tin tức
  Future<Map<String, dynamic>> fetchAllNewsAdmin({
    int page = 1,
    int limit = 20,
  }) async {
    final uri = Uri.parse('${ApiRoutes.adminNews}?page=$page&limit=$limit');
    final response = await http
        .get(
      uri,
      headers: await _headers(requireAuth: true),
    )
        .timeout(
      const Duration(seconds: 30),
      onTimeout: () => throw Exception('Timeout'),
    );

    if (response.statusCode == 200) {
      return jsonDecode(utf8.decode(response.bodyBytes));
    } else if (response.statusCode == 401) {
      throw Exception('Token hết hạn, vui lòng đăng nhập lại');
    } else if (response.statusCode == 403) {
      throw Exception('Bạn không có quyền Admin');
    }
    throw Exception('Tải danh sách tin tức thất bại');
  }

  /// Admin: Tạo tin tức mới (multiple images)
  Future<Map<String, dynamic>> createNews({
    required String title,
    required String content,
    required String summary,
    required List<File> imageFiles, // Thay đổi: danh sách file
    String author = 'Admin',
    String category = 'Tin tức',
    bool featured = false,
  }) async {
    if (imageFiles.isEmpty) {
      throw Exception('Vui lòng chọn ít nhất 1 ảnh');
    }

    // Validate tất cả file
    for (var file in imageFiles) {
      await _validateImageFile(file);
    }

    final request = http.MultipartRequest(
      'POST',
      Uri.parse(ApiRoutes.news),
    );

    // Header + token
    final headers = await _headers(requireAuth: true);
    request.headers.addAll(headers);

    // Fields
    request.fields.addAll({
      'title': title.trim(),
      'content': content,
      'summary': summary.trim(),
      'author': author.trim(),
      'category': category.trim(),
      'featured': featured.toString(),
    });

    // Thêm tất cả ảnh
    for (var imageFile in imageFiles) {
      final mimeType = _determineMimeType(imageFile.path);
      request.files.add(await http.MultipartFile.fromPath(
        'images', // Thay đổi: 'images' (plural)
        imageFile.path,
        contentType: MediaType.parse(mimeType),
      ));
    }

    final streamedResponse = await request.send().timeout(
      const Duration(seconds: 60),
      onTimeout: () => throw Exception('Upload timeout'),
    );
    final response = await http.Response.fromStream(streamedResponse);

    if (response.statusCode == 201) {
      return jsonDecode(response.body);
    } else if (response.statusCode == 401) {
      throw Exception('Token hết hạn, vui lòng đăng nhập lại');
    } else if (response.statusCode == 403) {
      throw Exception('Bạn không có quyền Admin');
    } else if (response.statusCode == 400) {
      final err = jsonDecode(response.body);
      throw Exception(err['message'] ?? 'Dữ liệu không hợp lệ');
    }
    throw Exception('Tạo tin tức thất bại (${response.statusCode})');
  }

  /// Admin: Cập nhật tin tức (multiple images)
  Future<Map<String, dynamic>> updateNews({
    required String newsId,
    required String title,
    required String content,
    required String summary,
    String author = 'Admin',
    String category = 'Tin tức',
    required bool featured,
    required bool isActive,
    List<File>? imageFiles, // Thay đổi: danh sách file
  }) async {
    if (imageFiles != null && imageFiles.isNotEmpty) {
      for (var file in imageFiles) {
        await _validateImageFile(file);
      }
    }

    final request = http.MultipartRequest(
      'PUT',
      Uri.parse(ApiRoutes.newsById(newsId)),
    );

    request.headers.addAll(await _headers(requireAuth: true));

    request.fields.addAll({
      'title': title.trim(),
      'content': content,
      'summary': summary.trim(),
      'author': author.trim(),
      'category': category.trim(),
      'featured': featured.toString(),
      'isActive': isActive.toString(),
    });

    // Thêm ảnh nếu có
    if (imageFiles != null && imageFiles.isNotEmpty) {
      for (var imageFile in imageFiles) {
        final mimeType = _determineMimeType(imageFile.path);
        request.files.add(await http.MultipartFile.fromPath(
          'images',
          imageFile.path,
          contentType: MediaType.parse(mimeType),
        ));
      }
    }

    final streamedResponse = await request.send().timeout(
      const Duration(seconds: 60),
      onTimeout: () => throw Exception('Upload timeout'),
    );
    final response = await http.Response.fromStream(streamedResponse);

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else if (response.statusCode == 401) {
      throw Exception('Token hết hạn');
    } else if (response.statusCode == 403) {
      throw Exception('Không có quyền Admin');
    } else if (response.statusCode == 404) {
      throw Exception('Tin tức không tồn tại');
    }
    throw Exception('Cập nhật thất bại (${response.statusCode})');
  }

  /// Admin: Xóa tin tức
  Future<void> deleteNews(String newsId) async {
    final uri = Uri.parse(ApiRoutes.newsById(newsId));
    final response = await http
        .delete(
      uri,
      headers: await _headers(requireAuth: true),
    )
        .timeout(
      const Duration(seconds: 30),
      onTimeout: () => throw Exception('Timeout'),
    );

    if (response.statusCode == 200) {
      return;
    } else if (response.statusCode == 401) {
      throw Exception('Token hết hạn, vui lòng đăng nhập lại');
    } else if (response.statusCode == 403) {
      throw Exception('Bạn không có quyền Admin');
    } else if (response.statusCode == 404) {
      throw Exception('Tin tức không tồn tại');
    }
    final err = jsonDecode(response.body);
    throw Exception(err['message'] ?? 'Xóa thất bại');
  }
}

File: App-RentHouse/routes/news.js
// routes/news.js
const express = require('express');
const router = express.Router();
const News = require('../models/news');
const SavedArticle = require('../models/savedArticle');
const multer = require('multer');
const cloudinary = require('../config/cloudinary');
const { CloudinaryStorage } = require('multer-storage-cloudinary');

// Cấu hình Cloudinary Storage cho multer (nhiều ảnh)
const storage = new CloudinaryStorage({
  cloudinary: cloudinary,
  params: {
    folder: 'news', // Thư mục trong Cloudinary
    allowed_formats: ['jpg', 'jpeg', 'png', 'webp'],
    transformation: [{ width: 1920, height: 1080, crop: 'limit' }], // Resize ảnh
  },
});

const upload = multer({
  storage,
  limits: { fileSize: 10 * 1024 * 1024 }, // 10MB
  fileFilter: (req, file, cb) => {
    const allowedExtensions = /\.(jpeg|jpg|png|webp)$/i;
    const extname = allowedExtensions.test(file.originalname);
    
    if (extname) {
      return cb(null, true);
    }
    cb(new Error('Chỉ chấp nhận file ảnh (JPEG, JPG, PNG, WebP)'));
  },
});

// Helper function: Xóa nhiều ảnh trên Cloudinary
const deleteCloudinaryImages = async (cloudinaryIds) => {
  if (!cloudinaryIds || cloudinaryIds.length === 0) {
    return [];
  }
  
  const results = [];
  for (const publicId of cloudinaryIds) {
    try {
      const result = await cloudinary.uploader.destroy(publicId);
      results.push({ publicId, result });
      console.log('Cloudinary delete:', publicId, result);
    } catch (error) {
      console.error('Error deleting from Cloudinary:', publicId, error);
      results.push({ publicId, error: error.message });
    }
  }
  return results;
};

// Middleware kiểm tra admin
const isAdmin = async (req, res, next) => {
  try {
    const token = req.headers.authorization?.split('Bearer ')[1];
    if (!token) return res.status(401).json({ message: 'Không có token' });

    const decoded = await require('firebase-admin').auth().verifyIdToken(token);
    const user = await require('../models/usermodel').findOne({ _id: decoded.uid });

    if (user?.role !== 'admin') {
      return res.status(403).json({ message: 'Bạn không có quyền admin' });
    }
    req.user = decoded;
    next();
  } catch (err) {
    console.error('Admin check error:', err);
    res.status(401).json({ message: 'Token không hợp lệ' });
  }
};

// Middleware kiểm tra user đã đăng nhập
const isAuth = async (req, res, next) => {
  try {
    const token = req.headers.authorization?.split('Bearer ')[1];
    if (!token) return res.status(401).json({ message: 'Không có token' });

    const decoded = await require('firebase-admin').auth().verifyIdToken(token);
    req.user = decoded;
    next();
  } catch (err) {
    console.error('Auth check error:', err);
    res.status(401).json({ message: 'Token không hợp lệ' });
  }
};

// ================== ROUTES CÔNG KHAI ==================

// Lấy tin tức công khai (active + featured)
router.get('/', async (req, res) => {
  try {
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 10;
    const skip = (page - 1) * limit;

    const news = await News.findActive()
      .skip(skip)
      .limit(limit);

    const total = await News.countDocuments({ isActive: true });

    res.json({
      data: news,
      pagination: {
        total,
        page,
        limit,
        pages: Math.ceil(total / limit),
      },
    });
  } catch (err) {
    console.error('Get news error:', err);
    res.status(500).json({ message: err.message });
  }
});

// Lấy tin tức nổi bật (featured)
router.get('/featured', async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 3;
    const news = await News.findFeatured(limit);
    res.json(news);
  } catch (err) {
    console.error('Get featured news error:', err);
    res.status(500).json({ message: err.message });
  }
});

// Lấy tin tức phổ biến (most views)
router.get('/popular', async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 10;
    const days = parseInt(req.query.days) || 30;
    const news = await News.findPopular(limit, days);
    res.json(news);
  } catch (err) {
    console.error('Get popular news error:', err);
    res.status(500).json({ message: err.message });
  }
});

// Lấy tin tức theo category
router.get('/category/:category', async (req, res) => {
  try {
    const { category } = req.params;
    const limit = parseInt(req.query.limit) || 10;
    const page = parseInt(req.query.page) || 1;
    const skip = (page - 1) * limit;

    const news = await News.findByCategory(category, limit + skip).skip(skip);
    const total = await News.countDocuments({ isActive: true, category });

    res.json({
      data: news,
      pagination: {
        total,
        page,
        limit,
        pages: Math.ceil(total / limit),
      },
    });
  } catch (err) {
    console.error('Get news by category error:', err);
    res.status(500).json({ message: err.message });
  }
});

// ================== ROUTES CÓ YÊU CẦU ĐĂNG NHẬP ==================

// Lấy tất cả tin tức đã lưu của user
router.get('/user/saved-articles', isAuth, async (req, res) => {
  try {
    const userId = req.user.uid;
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 10;
    const skip = (page - 1) * limit;

    console.log('Fetching saved articles for user:', userId);

    const savedArticles = await SavedArticle.find({ userId })
      .sort({ savedAt: -1 })
      .skip(skip)
      .limit(limit);

    const newsIds = savedArticles.map(item => item.newsId);
    const news = await News.find({ _id: { $in: newsIds }, isActive: true });

    const orderedNews = newsIds
      .map(id => news.find(n => n._id.toString() === id.toString()))
      .filter(Boolean);

    const total = await SavedArticle.countDocuments({ userId });

    res.json({
      data: orderedNews,
      pagination: {
        total,
        page,
        limit,
        pages: Math.ceil(total / limit),
      },
    });
  } catch (err) {
    console.error('Get saved articles error:', err);
    res.status(500).json({ message: err.message });
  }
});

// Lưu tin tức
router.post('/:newsId/save', isAuth, async (req, res) => {
  try {
    const { newsId } = req.params;
    const userId = req.user.uid;

    console.log('Saving article - User:', userId, 'News:', newsId);

    const news = await News.findById(newsId);
    if (!news) {
      return res.status(404).json({ message: 'Không tìm thấy tin tức' });
    }

    const existing = await SavedArticle.findOne({ userId, newsId });
    if (existing) {
      return res.status(400).json({ message: 'Bạn đã lưu tin tức này' });
    }

    const saved = new SavedArticle({ userId, newsId });
    await saved.save();

    console.log('Article saved successfully:', saved._id);
    res.status(201).json({
      message: 'Đã lưu tin tức',
      data: saved,
    });
  } catch (err) {
    console.error('Save article error:', err);
    res.status(500).json({ message: err.message });
  }
});

// Bỏ lưu tin tức
router.delete('/:newsId/unsave', isAuth, async (req, res) => {
  try {
    const { newsId } = req.params;
    const userId = req.user.uid;

    console.log('Unsaving article - User:', userId, 'News:', newsId);

    const result = await SavedArticle.findOneAndDelete({ userId, newsId });

    if (!result) {
      return res.status(404).json({ message: 'Không tìm thấy tin tức đã lưu' });
    }

    console.log('Article unsaved successfully');
    res.json({ message: 'Đã bỏ lưu tin tức' });
  } catch (err) {
    console.error('Unsave article error:', err);
    res.status(500).json({ message: err.message });
  }
});

// Kiểm tra tin tức có được lưu không
router.get('/:newsId/is-saved', isAuth, async (req, res) => {
  try {
    const { newsId } = req.params;
    const userId = req.user.uid;

    const saved = await SavedArticle.findOne({ userId, newsId });
    res.json({ isSaved: !!saved });
  } catch (err) {
    console.error('Check saved article error:', err);
    res.status(500).json({ message: err.message });
  }
});

// Lấy chi tiết tin tức (auto increment views)
router.get('/:id', async (req, res) => {
  try {
    const news = await News.incrementViews(req.params.id);

    if (!news || !news.isActive) {
      return res.status(404).json({ message: 'Không tìm thấy tin tức' });
    }

    // Lấy tin tức liên quan
    const related = await News.findRelated(news._id, news.category, 5);

    res.json({
      ...news.toJSON(),
      related,
    });
  } catch (err) {
    console.error('Get news detail error:', err);
    res.status(500).json({ message: err.message });
  }
});

// ================== ROUTES ADMIN ==================

// Admin: Lấy tất cả tin tức
router.get('/admin/all', isAdmin, async (req, res) => {
  try {
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 10;
    const skip = (page - 1) * limit;

    const news = await News.find()
      .sort({ createdAt: -1 })
      .skip(skip)
      .limit(limit);

    const total = await News.countDocuments();

    res.json({
      data: news,
      pagination: {
        total,
        page,
        limit,
        pages: Math.ceil(total / limit),
      },
    });
  } catch (err) {
    console.error('Get admin news error:', err);
    res.status(500).json({ message: err.message });
  }
});

// Admin: Thêm tin tức mới (multiple images)
router.post('/', isAdmin, upload.array('images', 10), async (req, res) => {
  let uploadedFiles = [];
  
  try {
    console.log('Create news - Body:', req.body);
    console.log('Create news - Files:', req.files?.length);

    const { title, content, summary, author, category, featured } = req.body;

    if (!req.files || req.files.length === 0) {
      return res.status(400).json({ message: 'Vui lòng upload ít nhất 1 ảnh' });
    }

    if (!title || !content) {
      // Xóa ảnh đã upload nếu validation fail
      uploadedFiles = req.files.map(f => f.filename);
      await deleteCloudinaryImages(uploadedFiles);
      return res.status(400).json({ message: 'Tiêu đề và nội dung không được bỏ trống' });
    }

    // Tạo mảng image data từ các file Cloudinary
    const images = req.files.map((file, index) => ({
      url: file.path, // URL từ Cloudinary
      cloudinaryId: file.filename, // Public ID
      order: index,
    }));

    const news = new News({
      title: title.trim(),
      content: content, // JSON Delta format từ Quill
      summary: summary ? summary.trim() : title.trim().substring(0, 100),
      images: images,
      imageUrl: images[0].url, // Ảnh đầu tiên (auto-sync bởi pre-save)
      imageUrls: images.map(img => img.url), // Auto-sync
      author: author ? author.trim() : 'Admin',
      category: category ? category.trim() : 'Tin tức',
      featured: featured === 'true' || featured === true,
      isActive: true,
    });

    await news.save();
    console.log('News created successfully:', news._id, 'Images:', images.length);
    res.status(201).json(news);
  } catch (err) {
    console.error('Create news error:', err);
    
    // Nếu lỗi, xóa tất cả ảnh vừa upload
    if (req.files?.length > 0) {
      uploadedFiles = req.files.map(f => f.filename);
      await deleteCloudinaryImages(uploadedFiles);
    }
    
    res.status(400).json({ message: err.message });
  }
});

// Admin: Cập nhật tin tức (multiple images)
router.put('/:id', isAdmin, upload.array('images', 10), async (req, res) => {
  let uploadedFiles = [];
  
  try {
    console.log('Update news - ID:', req.params.id);
    console.log('Update news - Body:', req.body);
    console.log('Update news - New files:', req.files?.length);

    const news = await News.findById(req.params.id);
    if (!news) {
      // Nếu không tìm thấy, xóa ảnh vừa upload
      if (req.files?.length > 0) {
        uploadedFiles = req.files.map(f => f.filename);
        await deleteCloudinaryImages(uploadedFiles);
      }
      return res.status(404).json({ message: 'Không tìm thấy tin tức' });
    }

    const updateData = {
      title: req.body.title ? req.body.title.trim() : news.title,
      content: req.body.content ? req.body.content : news.content,
      summary: req.body.summary ? req.body.summary.trim() : news.summary,
      author: req.body.author ? req.body.author.trim() : news.author,
      category: req.body.category ? req.body.category.trim() : news.category,
      featured: req.body.featured === 'true' || req.body.featured === true,
      isActive: req.body.isActive === 'true' || req.body.isActive === true,
      updatedAt: Date.now(),
    };

    // Lưu cloudinary IDs của ảnh cũ để xóa sau
    const oldCloudinaryIds = news.getCloudinaryDeleteInfo();

    // Nếu có ảnh mới, thay thế tất cả
    if (req.files && req.files.length > 0) {
      const newImages = req.files.map((file, index) => ({
        url: file.path,
        cloudinaryId: file.filename,
        order: index,
      }));
      
      updateData.images = newImages;
      // imageUrl và imageUrls sẽ auto-sync bởi pre-update middleware
    }

    const updated = await News.findByIdAndUpdate(
      req.params.id,
      updateData,
      { new: true, runValidators: true }
    );

    // Xóa ảnh cũ sau khi cập nhật thành công
    if (req.files && req.files.length > 0 && oldCloudinaryIds.length > 0) {
      await deleteCloudinaryImages(oldCloudinaryIds);
      console.log('Old images deleted from Cloudinary:', oldCloudinaryIds.length);
    }

    console.log('News updated successfully:', updated._id);
    res.json(updated);
  } catch (err) {
    console.error('Update news error:', err);
    
    // Nếu lỗi, xóa ảnh mới vừa upload
    if (req.files?.length > 0) {
      uploadedFiles = req.files.map(f => f.filename);
      await deleteCloudinaryImages(uploadedFiles);
    }
    
    res.status(400).json({ message: err.message });
  }
});

// Admin: Xóa tin tức
router.delete('/:id', isAdmin, async (req, res) => {
  try {
    console.log('Delete news - ID:', req.params.id);

    const news = await News.findById(req.params.id);
    if (!news) {
      return res.status(404).json({ message: 'Không tìm thấy tin tức' });
    }

    // Lấy cloudinary IDs để xóa
    const cloudinaryIds = news.getCloudinaryDeleteInfo();

    // Xóa document từ database
    await News.findByIdAndDelete(req.params.id);

    // Xóa ảnh trên Cloudinary
    if (cloudinaryIds.length > 0) {
      await deleteCloudinaryImages(cloudinaryIds);
      console.log('Images deleted from Cloudinary:', cloudinaryIds.length);
    }
    
    console.log('News deleted successfully:', req.params.id);
    res.json({ message: 'Xóa tin tức thành công' });
  } catch (err) {
    console.error('Delete news error:', err);
    res.status(500).json({ message: err.message });
  }
});

module.exports = router;